Lib/test/test_sqlite3/test_dbapi.py

```diff
@@ -523,6 +523,44 @@ def test_connection_init_good_isolation_levels(self):
                 with memory_database(isolation_level=level) as cx:
                     cx.execute("select 'ok'")
 
+    def test_connection_reinit(self):
+        db = ":memory:"
+        cx = sqlite.connect(db)
+        cx.text_factory = bytes
+        cx.row_factory = sqlite.Row
+        cu = cx.cursor()
+        cu.execute("create table foo (bar)")
+        cu.executemany("insert into foo (bar) values (?)",
+                       ((str(v),) for v in range(4)))
+        cu.execute("select bar from foo")
+
+        rows = [r for r in cu.fetchmany(2)]
+        self.assertTrue(all(isinstance(r, sqlite.Row) for r in rows))
+        self.assertEqual([r[0] for r in rows], [b"0", b"1"])
+
+        cx.__init__(db)
+        cx.execute("create table foo (bar)")
+        cx.executemany("insert into foo (bar) values (?)",
+                       ((v,) for v in ("a", "b", "c", "d")))
+
+        # This uses the old database, old row factory, but new text factory
```

Comment:
Why is this the expected behavior?

Reply:
I agree this is not currently obvious. This is why:

We're calling `fetchall` on the _cursor_. The active statement in the cursor is _"select bar from foo"_. We've already read the two first values from it (line 537). Since line 537, we've "reloaded" the connection. When we call `fetchall` again, the cursor still uses the "old" statement (AKA "using the old database" because the `sqlite3_stmt` pointer preserves that connection). SQLite correctly returns the remaining rows from the query into the previous database. The row factory was inherited from the (old) connection. It should stay the same, since we didn't touch the cursor. The text factory is "read" directly from the connection pointer, so this will change when we reboot the connection.